### PR TITLE
dma: switch to native Zephyr API for dw-dma & HDA

### PR DESCRIPTION
**Pull request by abonislawski**
This patch will add dw-dma wrapper for zephyr driver

Opens: 
- sof dma ops names: dma_channel_put_sof? dw_dma_channel_put_zephyr?
- zephyr/CMakeLists.txt: dma-zephyr.c should be restricted for a specific platform/s?

### CODE REVIEW

**Comment by teburd** [comment]
This seems like a nice way to adapt the SOF <-> Zephyr APIs

File: src/drivers/dw/dma-zephyr.c [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+			return ret;
+	}
+
+	if (next.status == DMA_CB_STATUS_END)
+		dw_dma_stop_zephyr(channel);
+
+	ret = dma_reload(channel->dma->z_drv, channel->index, 0, 0, bytes);
+	if (ret < 0)
+		tr_dbg(&dwdma_tr, "dw_dma_copy(): dma %d channel %d fail",
+		       channel->dma->plat_data.id, channel->index);
```
Comment (lyakh) [inline src/drivers/dw/dma-zephyr.c:234]: just a `tr_dbg()`?

File: src/drivers/dw/dma-zephyr.c [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+	tr_dbg(&dwdma_tr, "dw_dma_get_data_size(): dma %d channel %d get data size",
+	       channel->dma->plat_data.id, channel->index);
+
+	ret = dma_get_status(channel->dma->z_drv, channel->index, &stat);
+	*avail = stat.pending_length;
+	*free = stat.free;
+
+	if (ret < 0) {
+		tr_err(&dwdma_tr, "dw_dma_get_data_size(): xrun detected");
+		return -ENODATA;
```
Comment (lyakh) [inline src/drivers/dw/dma-zephyr.c:301]: not propagating?

**Review by lyakh — CHANGES_REQUESTED** [review]

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Needs to be more generic and not per DMA driver.

**Comment by sys-pt1s** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
@abonislawski any update ?
My expectation is that for each DMA API call we have today we would map current users onto the Zephyr version of that call and wrap the SOF version (for existing xtos users).
e.g. for dma_stop()
```
 grep -rn " dma_stop(" --include=*.c
src/trace/dma-trace.c:375:		err = dma_stop(d->dc.chan);
src/audio/host.c:604:		ret = dma_stop(hd->chan);
src/audio/dai.c:782:		ret = dma_stop(dd->chan);
src/audio/dai.c:786:		ret = dma_stop(dd->chan);
src/probe/probe.c:175:	err = dma_stop(dma->dc.chan);
src/probe/probe.c:177:		tr_err(&pr_tr, "probe_dma_deinit(): dma_stop() failed");
```
These would be mapped to the native Zephyr API call i.e. all user would be converted to use

int dma_stop(const struct [device](https://docs.zephyrproject.org/3.0.0/reference/drivers/index.html#c.device) *dev, uint32_t channel)[¶](https://docs.zephyrproject.org/3.0.0/reference/peripherals/dma.html#c.dma_stop)

We would need to convert the existing SOF dma calls to the Zephyr APIs as the first part of this work and then add in support for linking with the Zephyr implementation of this API. 

@juimonen same applies to SSP and other DAI drivers.

**Comment by abonislawski** [comment]
@lgirdwood this approach will require to switch all dma or all dai drivers once. With sof api -> zephyr api wrapper we can merge and test each driver separately, it is also possible to use for example one dai driver from sof and one from zephyr.
Can we go this way until all drivers are ready and then in separate step switch dma/dai api to native zephyr?

**Comment by lgirdwood** [comment]
> @lgirdwood this approach will require to switch all dma or all dai drivers once. With sof api -> zephyr api wrapper we can merge and test each driver separately, it is also possible to use for example one dai driver from sof and one from zephyr. Can we go this way until all drivers are ready and then in separate final step switch dma/dai api to native zephyr?

I want to avoid having an intermediate stage to develop and validate, I think it's fine to transition in steps with a Kconfig option "Use Zephyr drivers" with default value n. This means that all the local xtos based driver will be used unless this Kconfig is "y". 
i.e. stages with Zephyr drivers enabled.
1)  HDA DMA and HDA/HDMI audio - only HDA and trace can be tested.
2) DW DMA and DMIC, SSP - DMIC and SSP can also be tested.
3) Remaining Intel DAIs - all features now enabled.

File: src/audio/host_zephyr.c [hunk 3]
```diff
@@ -0,0 +1,1135 @@
...
+}
+
+/* configure the DMA params and descriptors for host buffer IO */
+static int host_params(struct comp_dev *dev,
+		       struct sof_ipc_stream_params *params)
+{
+	struct host_data *hd = comp_get_drvdata(dev);
+	struct dma_sg_config *config = &hd->config;
+	struct dma_sg_elem *sg_elem;
+	struct dma_config *dma_cfg = &hd->z_config;
```
Comment (abonislawski) [inline src/audio/host_zephyr.c:768]: The main issue is dma_config, Im still using old config and translating it to the dma_config at the end, which is obviously wrong but the old config contain some fields which are not available in the new config, like is_scheduling_source, not sure if we can ignore such fields?

File: src/audio/host_zephyr.c [hunk 4]
```diff
@@ -0,0 +1,1135 @@
...
+
+	/* host params always installed by pipeline IPC */
+	hd->host_size = params->buffer.size;
+	hd->stream_tag = params->stream_tag;
+	hd->no_stream_position = params->no_stream_position;
+	hd->host_period_bytes = params->host_period_bytes;
+	hd->cont_update_posn = params->cont_update_posn;
+
+	/* retrieve DMA buffer address alignment */
+	err = dma_get_attribute(hd->dma, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
```
Comment (abonislawski) [inline src/audio/host_zephyr.c:802]: dma_get_attrubite is still missing in new version, Zephyr API does not support this.
Not sure if we can use here defines directly?

**Review by abonislawski — COMMENTED** [review]
@lgirdwood please take a look, added early version but few things still needs to be changed

added host_zephyr.c and dai_zephyr.c, so both HDA and DW-DMA (similar changes)
sof dma lib not divided yet

Please ignore _sof functions, it will be reverted later

File: src/audio/dai.c [hunk 5]
```diff
@@ -218,7 +218,7 @@ static void dai_free(struct comp_dev *dev)
 
 	if (dd->chan) {
 		notifier_unregister(dev, dd->chan, NOTIFIER_ID_DMA_COPY);
-		dma_channel_put(dd->chan);
+		dma_channel_put_sof(dd->chan);
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/audio/dai.c:223]: with native api _sof functions will not be needed so I will remove them

File: src/include/sof/lib/dai.h [hunk 6]
```diff
@@ -157,6 +157,7 @@ struct dai_data {
 	struct dma_chan_data *chan;
 	uint32_t stream_id;
 	struct dma_sg_config config;
+	struct dma_config z_config;
```
Comment (kv2019i) [inline src/include/sof/lib/dai.h:160]: Should this be ifdeffed for ZEPHYR?

**Review by kv2019i — CHANGES_REQUESTED** [review]
Looks good for the most part. A few minor comments, but no issue with the overall approach.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
Both this and #5731 are/were abstracting at the wrong point. The abstraction points for DMA are ```src/audio/dai.c``` and ```src/audio/host.c```.
This needs to be mechanically changed to
1) create a dai-zephyr.c (starts to use native Zephyr APIs) and copy dai.c to dai-legacy.c (this last file is exact copy of dai.c today plus any legacy dai logic from lib/dai.c).
2) same as 1) for host.c
Good for call if needed.

**Comment by lgirdwood** [comment]
> Both this and #5731 are/were abstracting at the wrong point. The abstraction points for DMA are `src/audio/dai.c` and `src/audio/host.c`. This needs to be mechanically changed to
> 
> 1. create a dai-zephyr.c (starts to use native Zephyr APIs) and copy dai.c to dai-legacy.c (this last file is exact copy of dai.c today plus any legacy dai logic from lib/dai.c).
> 2. same as 1) for host.c
>    Good for call if needed.

One more thing, lets not use ```_sof``` suffixes for the legacy APIs. We can use ```_legacy``` instead and this would also align with the file name.

**Comment by dbaluta** [comment]
@abonislawski , maybe this could be useful. 

FYI: https://github.com/zephyrproject-rtos/zephyr/pull/44999/commits/2035f3aa2a849a61404df0444336d9a46523b165

**Comment by abonislawski** [comment]
> Both this and #5731 are/were abstracting at the wrong point. The abstraction points for DMA are `src/audio/dai.c` and `src/audio/host.c`. This needs to be mechanically changed to
> 
> 1. create a dai-zephyr.c (starts to use native Zephyr APIs) and copy dai.c to dai-legacy.c (this last file is exact copy of dai.c today plus any legacy dai logic from lib/dai.c).
> 2. same as 1) for host.c
>    Good for call if needed.

@lgirdwood please check carefully, this is how it works now with native zephyr api in src/audio/dai_zephyr.c and src/audio/host_zephyr.c, I can also rename them to dai.c/host.c and old version to dai-legacy.c/host-legacy.c

**Comment by lgirdwood** [comment]
> @lgirdwood please check carefully, this is how it works now with native zephyr api in src/audio/dai_zephyr.c and src/audio/host_zephyr.c, I can also rename them to dai.c/host.c and old version to dai-legacy.c/host-legacy.c

Yes please - we really need to make it obvious what files are Zephyr and what files are legacy. Also please use the ```_legacy```  suffix. This gives us a clear dividing line.

File: src/include/sof/lib/dai.h (hunk 6 above)
Comment (abonislawski) [reply src/include/sof/lib/dai.h]: sure, why not :)

**Comment by abonislawski** [comment]
@lgirdwood updated with new legacy/zephyr suffix, also added CONFIG_ZEPHYR_NATIVE_DRIVERS as you adviced previously so we can enable it later when everything is ready along with dai PR

**Review by lgirdwood — APPROVED** [review]
This is a good first step toward full native Zephyr. Next things (in subsequent PRs)  are
1) Reduce other SOF native RTOS APIs in host-zephyr.c and dai-zephyr.c and use the Zephyr versions, i.e for PM, notifiers memory.
2) Fix the DAI, DMA init to have const tables of DT name strings and topology IDs so that teh CC does more of the init and not the runtime code.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by abonislawski** [comment]
@lyakh I will fix most of your comments today but please note that this PR is not about dai&host refactor

File: src/audio/dai-zephyr.c [hunk 7]
```diff
@@ -0,0 +1,1178 @@
...
+	if (dd->xrun) {
+		/* make sure we only playback silence during an XRUN */
+		if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
+			/* fill buffer with silence */
+			buffer_zero(dd->dma_buffer);
+
+		return;
+	}
+
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/audio/dai-zephyr.c:114]: This is not a brand new dai component, just modified a little dma things, flows are the same

File: src/audio/dai-zephyr.c [hunk 8]
```diff
@@ -0,0 +1,1178 @@
...
+			/* The counter has reached 0, which means
+			 * all DAIs have received the same trigger command
+			 * and we may begin the actual trigger process
+			 * synchronously.
+			 */
+
+			irq_local_disable(irq_flags);
+			notifier_event(group, NOTIFIER_ID_DAI_TRIGGER,
+				       BIT(cpu_get_id()), NULL, 0);
+			irq_local_enable(irq_flags);
```
[the remark this answers is not in the document]
Comment (abonislawski) [reply src/audio/dai-zephyr.c:951]: I can removed this but this is just our original code, no changes in this PR

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**